algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -12,258 +12,321 @@
-                    *alpha,
-                    beta,
-                    v_H_i_alpha_v_H_i_beta,
-                    largest_non_zero_domain_size,
-                    &circuit.fft_precomputation,
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-
-            let label_g_c = witness_label(circuit.id, "g_c", 0);
-            pool.add_job(move || {
-                let result = Self::matrix_sumcheck_helper(
-                    label_g_c,
-                    circuit_state.non_zero_c_domain,
-                    &circuit.c_arith,
-                    *alpha,
-                    beta,
-                    v_H_i_alpha_v_H_i_beta,
-                    largest_non_zero_domain_size,
-                    &circuit.fft_precomputation,
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-        }
+        let zk_bound = Self::zk_bound();
 
-        let mut sums = Vec::with_capacity(state.circuit_specific_states.len());
-        let mut gs = BTreeMap::new();
-        for ((circuit_a, (sum_a, lhs_a, g_a)), (circuit_b, (sum_b, lhs_b, g_b)), (circuit_c, (sum_c, lhs_c, g_c))) in
-            pool.execute_all().into_iter().tuples()
+        let max_variable_domain = state.max_variable_domain;
+
+        let verifier::FirstMessage { batch_combiners } = verifier_message;
+        let verifier::SecondMessage { alpha, eta_b, eta_c } = verifier_second_message;
+
+        let assignments = Self::calc_assignments(&mut state)?;
+        let matrix_transposes = Self::calc_matrix_transpose(&mut state)?;
+
+        let (h_1, x_g_1_sum, msg) = Self::calc_lineval_sumcheck_witness(
+            &mut state,
+            batch_combiners,
+            assignments,
+            matrix_transposes,
+            alpha,
+            eta_b,
+            eta_c,
+        )?;
+
+        #[cfg(debug_assertions)]
         {
-            assert_eq!(circuit_a, circuit_b);
-            assert_eq!(circuit_a, circuit_c);
-            let matrix_sum = prover::message::MatrixSums { sum_a, sum_b, sum_c };
-            sums.push(matrix_sum);
-            state.circuit_specific_states.get_mut(circuit_a).unwrap().lhs_polynomials = Some([lhs_a, lhs_b, lhs_c]);
-            let matrix_gs = prover::MatrixGs { g_a, g_b, g_c };
-            gs.insert(circuit_a.id, matrix_gs);
+            let mut sumcheck_lhs = h_1.mul_by_vanishing_poly(max_variable_domain);
+            sumcheck_lhs += &x_g_1_sum;
+            debug_assert!(
+                sumcheck_lhs.evaluate_over_domain_by_ref(max_variable_domain).evaluations.into_iter().sum::<F>()
+                    == msg.sum(batch_combiners, *eta_b, *eta_c)
+            );
         }
 
-        let msg = prover::ThirdMessage { sums };
-        let oracles = prover::ThirdOracles { gs };
+        let g_1 = DensePolynomial::from_coefficients_slice(&x_g_1_sum.coeffs[1..]);
+
+        drop(x_g_1_sum);
```

Comment:
a comment with the reason for this `drop` would be useful

Reply:
For context, this was already there and just as secret/sensitive as a lot of other values we don't `drop`... So I guess it is defense in depth and clarity for the dev: ab096deab

Reply:
It's there to reduce the memory usage: we can drop things once they're no longer used in the rest of the algorithm.